Source/santad/EventProviders/SNTEndpointSecurityManager.mm

```diff
@@ -0,0 +1,304 @@
+          strncpy(sm.newpath, m->event.rename.destination.new_path.filename.data, m->event.rename.destination.new_path.filename.length);
+          sm.newpath[m->event.rename.destination.new_path.filename.length] = '\0';
+          break;
+        case ES_DESTINATION_TYPE_EXISTING_FILE:
+          strncpy(sm.newpath, m->event.rename.destination.existing_file->path.data, m->event.rename.destination.existing_file->path.length);
+          sm.newpath[m->event.rename.destination.existing_file->path.length] = '\0';
+          break;
+      }
+
+      callback = self.logCallback;
+      audit_token = m->process->audit_token;
+    }
+    default:
+      break;
+  }
+
+  if (self.prefixTree->HasPrefix(sm.path)) {
+    return;
+  }
+
+  if (callback) {
+    sm.uid = audit_token_to_ruid(audit_token);
+    sm.gid = audit_token_to_rgid(audit_token);
+    sm.pid = audit_token_to_pid(audit_token);
+    proc_name(sm.pid, sm.pname, 1024);
+    callback(sm);
+  }
+}
+
+- (void)listenForDecisionRequests:(void (^)(santa_message_t))callback API_AVAILABLE(macos(10.15)) {
+  while (!self.connectionEstablished) usleep(100000); // 100ms
+
+  // Listen for exec auth messages.
+  self.decisionCallback = callback;
+  es_event_type_t events[] = { ES_EVENT_TYPE_AUTH_EXEC };
+  es_return_t sret = es_subscribe(self.client, events, 1);
+  if (sret != ES_RETURN_SUCCESS) LOGE(@"Unable to subscribe ES_EVENT_TYPE_AUTH_EXEC: %d", sret);
+
+  // There's a gap between creating a client and subscribing to events. Creating the client
+  // triggers a cache flush automatically but any events that happen in this gap could be allowed
+  // and cached, so we force the cache to flush again.
+  [self flushCacheNonRootOnly:YES];
+}
+
+- (void)listenForLogRequests:(void (^)(santa_message_t))callback API_AVAILABLE(macos(10.15)) {
+  while (!self.connectionEstablished) usleep(100000); // 100ms
+
+  // Listen for exec notify messages.
+  self.logCallback = callback;
+  es_event_type_t events[] = {
+    ES_EVENT_TYPE_NOTIFY_EXEC,
+    ES_EVENT_TYPE_NOTIFY_CLOSE,
+    ES_EVENT_TYPE_NOTIFY_TRUNCATE,
+    ES_EVENT_TYPE_NOTIFY_LINK,
+    ES_EVENT_TYPE_NOTIFY_RENAME,
+    ES_EVENT_TYPE_NOTIFY_UNLINK,
+  };
+  es_return_t sret = es_subscribe(self.client, events, sizeof(events) / sizeof(es_event_type_t));
+  if (sret != ES_RETURN_SUCCESS) LOGE(@"Unable to subscribe ES_EVENT_TYPE_NOTIFY_EXEC: %d", sret);
+}
+
+- (int)postAction:(santa_action_t)action forMessage:(santa_message_t)sm
+    API_AVAILABLE(macos(10.15)) {
+  es_respond_result_t ret;
+  switch (action) {
+    case ACTION_RESPOND_ALLOW:
+      ret = es_respond_auth_result(self.client, (es_message_t *)sm.es_message,
+                                   ES_AUTH_RESULT_ALLOW, true);
+      break;
+    case ACTION_RESPOND_DENY:
+      ret = es_respond_auth_result(self.client, (es_message_t *)sm.es_message,
+                                   ES_AUTH_RESULT_DENY, false);
+      break;
+    default:
+      return ES_RESPOND_RESULT_ERR_INVALID_ARGUMENT;
```

Comment:
So we don't leak accidentally leak sm.es_message.
ret = ES_RESPOND_RESULT_ERR_INVALID_ARGUMENT

Reply:
Done